dbms/src/Storages/Page/V3/PageDirectory.cpp

```diff
@@ -904,10 +939,31 @@ PageId PageDirectory::getMaxId(NamespaceId ns_id) const
         // iter is not at the beginning and mvcc_table_directory is not empty,
         // so iter-- must be a valid iterator, and it's the largest page id which is smaller than the target page id.
         iter--;
-        if (iter->first.high == ns_id)
-            return iter->first.low;
-        else
-            return 0;
+
+        do
+        {
+            // Can't find any entries in current ns_id
+            if (iter->first.high != ns_id)
+            {
+                break;
+            }
+
+            // Check and return whether this id is visible, otherwise continue to check the previous one.
+            if (iter->second->isVisible(UINT64_MAX - 1))
+            {
+                return iter->first.low;
+            }
+
+            // Current entry/ref/external is deleted and there are no entries before it.
+            if (iter == mvcc_table_directory.begin())
+            {
+                break;
+            }
+
+            iter--;
+        } while (true);
```

Comment:
These changes revert https://github.com/pingcap/tiflash/pull/4845 but use `isVisible` to check whether we can return the id in getMaxId /cc @jiaqizho @lidezhu